### PR TITLE
Implement missing Foundation utility functions

### LINKED ISSUE #2678: Implement NSStringFromProtocol, NSProtocolFromString, and NSGetSizeAndAlignment

### PR DESCRIPTION
**Pull request by aballway**
Implements NSStringFromProtocol, NSProtocolFromString, and NSGetSizeAndAlignment

Fixes #2678

### CODE REVIEW

File: tests/UnitTests/Foundation/FoundationTests.mm [hunk 1]
```diff
@@ -1253,3 +1253,27 @@ @implementation TestKVOObject2
...
+    EXPECT_OBJCEQ(@"Protocol2", NSStringFromProtocol(@protocol(Protocol2)));
+    EXPECT_OBJCEQ(@"nil", NSStringFromProtocol(nil));
+}
+
+TEST(Foundation, NSProtocolFromString) {
+    EXPECT_EQ(@protocol(Protocol1), NSProtocolFromString(@"Protocol1"));
+    EXPECT_EQ(@protocol(Protocol2), NSProtocolFromString(@"Protocol2"));
+    EXPECT_EQ(nil, NSProtocolFromString(nil));
+    EXPECT_EQ(nil,
+              NSProtocolFromString(@"NonexistentProtocolOrAtLeastOneThatShouldNotEverEVERExistBecauseOfTheObnoxiouslyLongNameAsWellAsI"
```
Comment (DHowett-MSFT) [inline tests/UnitTests/Foundation/FoundationTests.mm]: :rofl:

File: tests/UnitTests/Foundation/WindowsOnly/FoundationInternalTests.mm [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+class TypeSizeAlignmentTest : public ::testing::TestWithParam<const char*> {};
+
+TEST_P(TypeSizeAlignmentTest, NSGetSizeAndAlignment) {
+    NSUInteger size = 0;
+    NSUInteger align = 0;
+    const char* type = GetParam();
+    ASSERT_NO_THROW(NSGetSizeAndAlignment(type, &size, &align));
+    EXPECT_EQ(objc_sizeof_type(type), size);
```
Comment (DHowett-MSFT) [inline tests/UnitTests/Foundation/WindowsOnly/FoundationInternalTests.mm]: if only you could do a type-parameterized test to test the `sizeof`/`alignof` for the actual backing type...

File: Frameworks/Foundation/Utils.mm [hunk 3]
```diff
@@ -55,28 +56,30 @@ SEL NSSelectorFromString(NSString* str) {
...
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
 NSString* NSStringFromProtocol(Protocol* proto) {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return proto ? [NSString stringWithCString:protocol_getName(proto)] : @"nil";
```
Comment (msft-Jeyaram) [inline Frameworks/Foundation/Utils.mm]: LOL

File: tests/UnitTests/Foundation/WindowsOnly/FoundationInternalTests.mm (hunk 2 above)
Comment (aballway) [inline tests/UnitTests/Foundation/WindowsOnly/FoundationInternalTests.mm]: Unfortunately we don't have that shimmed for TAEF, and it was not worth it for this 😅

**Pull request merged**